tianshou/policy/modelfree/bdq.py

```diff
@@ -0,0 +1,161 @@
+        target_q = np.repeat(target_q[..., None], self.max_action_num, axis=-1)
+
+        batch.returns = to_torch_as(target_q, target_q_torch)
+        if hasattr(batch, "weight"):  # prio buffer update
+            batch.weight = to_torch_as(batch.weight, target_q_torch)
+        return batch
+
+    def process_fn(
+        self, batch: Batch, buffer: ReplayBuffer, indices: np.ndarray
+    ) -> Batch:
+        """Compute the 1-step return for BDQ targets."""
+        batch = self._compute_return(batch, buffer, indices)
+        return batch
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[Dict, Batch, np.ndarray]] = None,
+        model: str = "model",
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data.
+
+        :param float eps: in [0, 1], for epsilon-greedy exploration method.
+
+        :return: A :class:`~tianshou.data.Batch` which has 3 keys:
+
+            * ``act`` the action.
+            * ``logits`` the network's raw output.
+            * ``state`` the hidden state.
+
+        .. seealso::
+
+            Please refer to :meth:`~tianshou.policy.BasePolicy.forward` for
+            more detailed explanation.
+        """
+        model = getattr(self, model)
+        obs = batch[input]
+        obs_next = obs.obs if hasattr(obs, "obs") else obs
+        logits, hidden = model(obs_next, state=state, info=batch.info)
+        act = to_numpy(logits.max(dim=-1)[1])
+        return Batch(logits=logits, act=act, state=hidden)
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        if self._target and self._iter % self._freq == 0:
+            self.sync_weight()
+        self.optim.zero_grad()
+        weight = batch.pop("weight", 1.0)
+        act = torch.tensor(batch.act).to(batch.returns.get_device())
+        q = self(batch).logits
+        act_mask = torch.zeros_like(q)
+        act_mask = act_mask.scatter_(-1, act.unsqueeze(-1), 1)
+        act_q = q * act_mask
+        returns = batch.returns
+        returns = returns * act_mask
+        td_error = (returns - act_q)
+        loss = (td_error.pow(2).sum(-1).mean(-1) * weight).mean()
+        # batch.weight = td_error.sum(-1).sum(-1)  # prio-buffer
+        loss.backward()
+        self.optim.step()
+        self._iter += 1
+        return {"loss": loss.item()}
+
+    def exploration_noise(self, act: Union[np.ndarray, Batch],
+                          batch: Batch) -> Union[np.ndarray, Batch]:
+        if isinstance(act, np.ndarray) and not np.isclose(self.eps, 0.0):
+            bsz = len(act)
+            rand_mask = np.random.rand(bsz) < self.eps
+            rand_act = np.random.randint(
+                0, self.max_action_num, (bsz, act.shape[-1])
+            )  # [0, 1]
+            if hasattr(batch.obs, "mask"):
+                rand_act += batch.obs.mask
+            act[rand_mask] = rand_act[rand_mask]
+        return act
```

Comment:
why keep this?

Reply:
@Trinkle23897 All of the above have subtle but important changes w.r.t the classic Q-learning. To sum up, in the classic Q-learning the Q-model has an output of the shape (bcz, total_num_actions), while for the Branching Q-model it has a shape (bcz, action_dim, num_action_per_dim). This is how it scales linearly w.r.t the number of action space dimensions.

Reply:
Is it possible to extend the existing functions to support this feature?

Reply:
The only way I see to extend the existing functions is by doing a lot of "if"s and "else"s which is not elegant. Also we discussed previously that it's better to leave DQN as is since it is such a classic algorithm. To build upon it it's better to inherit and then modify, which is what I did.